src/Symfony/Component/HttpFoundation/Response.php

```diff
@@ -326,30 +331,64 @@ public function prepare(Request $request): static
     /**
      * Sends HTTP headers.
      *
+     * @param null|positive-int $statusCode The status code to use. Override the statusCode property if set and not null.
+     *
      * @return $this
      */
-    public function sendHeaders(): static
+    public function sendHeaders(/* ?int $statusCode = null */): static
     {
         // headers have already been sent by the developer
         if (headers_sent()) {
             return $this;
         }
 
+        $statusCode = \func_num_args() > 0 ? func_get_arg(0) : null;
+        $informationalResponse = $statusCode >= 100 && $statusCode < 200;
+        if ($informationalResponse && !\function_exists('headers_send')) {
+            // skip informational responses if not supported by the SAPI
+            return $this;
+        }
+
         // headers
         foreach ($this->headers->allPreserveCaseWithoutCookies() as $name => $values) {
+            $previousValues = $this->sentHeaders[$name] ?? null;
+            if ($previousValues === $values) {
+                // Header already sent in a previous response, it will be automatically copied in this response by PHP
```

Comment:
Yes FrankenPHP does that automatically (it delegates to Go which does it automatically too). I don't think it's necessary to introduce more complexity yet because FrankenPHP is currently the only SAPI supporting this feature, and if new SAPIs add it they will likely work the same way (if that's not the case, we'll still be able to patch at that time).

Reply:
As I see it, this behavior from FrakenPHP is introducing complexity on our side. I'd better have it *not* keep early headers in memory and rely on the app to resend its headers in the real response.
I feel like the current behavior is also not future-proof and is coupled to 103. But what if another status code decides to *not* require repeating some headers?

Reply:
Go and FrankenPHP do it only for 103, so it's future proof. We should probably do it only for 103 here too.

It definitely introduces some complexity on Symfony-side, but it makes it more easy for all users using PHP directly. Also, it's a bit faster.

Reply:
> We should probably do it only for 103 here too

Definitely! I would prefer removing this special case for 103. If it's still doable, please consider my modest request :)
The complexity this requires on our side says why, no need for more arguments.

Reply:
I implemented in FrankenPHP what we discussed together in https://github.com/php/php-src/pull/7025#issuecomment-848015799 :)

>  I'd better have it not keep early headers in memory

FrankenPHP just delegates to Go... which does that automatically. To bypass this Go behavior we'll probably have to introduce the complexity in FrankenPHP, and this will come at the cost of a performance penalty.

> The complexity this requires on our side says why, no need for more arguments.

As we discussed on the original issue on php-src, this behavior is implied by the spec and not having it implemented by the SAPI will make 103 cumbersome (and error-prone) to use in plain PHP (without Symfony).

I agree that this complexity in Symfony is not ideal, but it's due to how Symfony is designed, and this will be hidden from the end-user anyway.

> Definitely! I would prefer removing this special case for 103. 

We do this exception for 103 in Go (and FrankenPHP) because this behavior doesn't apply to all 1XX codes, but only to 103 (it is described in the Early Hints RFC).

I'll update the code to do that only for 103 (it was a mistake to not do it immediately).